File: MIDTERM-LAB-2/uphsd-cs-nagpaton-francescalouise/linear-algebra/assignment-01/DeterminantSolver.java
/**
 * =====================================================
 * Student Name    : Nagpaton, Francesca Louise May G.
 * Course          : Math 101 — Linear Algebra
 * Assignment      : Programming Assignment 1 — 3x3 Matrix Determinant Solver
 * School          : University of Perpetual Help System DALTA, Molino Campus
 * Date            : March 19, 2026
 * GitHub Repo     : https://github.com/honeyshey-ops/Prog2-9302-AY225-NAGPATON
 *
 * Description:
 *   This program computes the determinant of a hardcoded 3x3 matrix assigned
 *   to Nagpaton, Francesca Louise May G. for Math 101. The solution is computed
 *   using cofactor expansion along the first row. Each intermediate step
 *   (2x2 minor, cofactor term, running sum) is printed to the console in a
 *   readable, step-by-step format so the full solution process is transparent.
 * =====================================================
 */
public class DeterminantSolver {

    // ── SECTION 1: Matrix Declaration ───────────────────────────────────
    // This is the 3x3 matrix assigned to student #23, Nagpaton Francesca Louise May G.
    // It is stored as a 2D integer array in row-major order (rows first, then columns).
    // The values come directly from the assignment sheet and must not be changed.
    static int[][] matrix = {
        { 1, 3, 5 },   // Row 1 of assigned matrix
        { 2, 4, 6 },   // Row 2 of assigned matrix
        { 3, 5, 7 }    // Row 3 of assigned matrix
    };

    // ── SECTION 2: 2×2 Determinant Helper ───────────────────────────────
    // This method takes four integers representing a 2x2 matrix arranged as:
    //   | a  b |
    //   | c  d |
    // and returns the determinant using the formula: (a * d) - (b * c).
    // It is called once per minor during the cofactor expansion.
    static int computeMinor(int a, int b, int c, int d) {
        // Multiply the main diagonal and subtract the anti-diagonal product
        return (a * d) - (b * c);
    }

    // ── SECTION 3: Matrix Printer ────────────────────────────────────────
    // This method neatly displays the 3x3 matrix inside a box border so it
    // looks like a proper mathematical matrix in the console output.
    // It loops through each row and prints the three values with padding.
    static void printMatrix(int[][] m) {
        System.out.println("┌                  ┐");
        for (int[] row : m) {
            System.out.printf("│  %3d  %3d  %3d   │%n", row[0], row[1], row[2]);
        }
        System.out.println("└                  ┘");
    }

    // ── SECTION 4: Step-by-Step Determinant Solver ──────────────────────
    // This is the core method that does the full cofactor expansion and
    // prints every step of the process. The steps are:
    //   1. Show the assigned matrix
    //   2. Compute each of the three 2x2 minors (M11, M12, M13)
    //   3. Print the arithmetic used to get each minor value
    //   4. Calculate each signed cofactor term (C11, C12, C13)
    //   5. Sum the cofactors and display the final determinant
    //   6. Warn if the determinant is zero (singular matrix)
    static void solveDeterminant(int[][] m) {

        // Print the section header and display the matrix
        System.out.println("=".repeat(52));
        System.out.println("  3x3 MATRIX DETERMINANT SOLVER");
        System.out.println("  Student: Nagpaton, Francesca Louise May G.");
        System.out.println("  Assigned Matrix:");
        System.out.println("=".repeat(52));
        printMatrix(m);
        System.out.println("=".repeat(52));
        System.out.println();
        System.out.println("Expanding along Row 1 (cofactor expansion):");
        System.out.println();

        // ── Step 1: Compute minor M₁₁ ──
        // To get M11, cross out row 0 and column 0.
        // The four remaining elements form the 2x2 sub-matrix used here.
        int minor11 = computeMinor(m[1][1], m[1][2], m[2][1], m[2][2]);
        System.out.printf(
            "  Step 1 Minor M: det([%d,%d],[%d,%d]) = (%d*%d) - (%d*%d) = %d - %d = %d%n",
            m[1][1], m[1][2], m[2][1], m[2][2],
            m[1][1], m[2][2], m[1][2], m[2][1],
            m[1][1] * m[2][2], m[1][2] * m[2][1],
            minor11
        );

        // ── Step 2: Compute minor M₁₂ ──
        // Cross out row 0 and column 1 to isolate the 2x2 sub-matrix below.
        int minor12 = computeMinor(m[1][0], m[1][2], m[2][0], m[2][2]);
        System.out.printf(
            "  Step 2 Minor M: det([%d,%d],[%d,%d]) = (%d*%d) - (%d*%d) = %d - %d = %d%n",
            m[1][0], m[1][2], m[2][0], m[2][2],
            m[1][0], m[2][2], m[1][2], m[2][0],
            m[1][0] * m[2][2], m[1][2] * m[2][0],
            minor12
        );

        // ── Step 3: Compute minor M₁₃ ──
        // Cross out row 0 and column 2 to get the last 2x2 sub-matrix.
        int minor13 = computeMinor(m[1][0], m[1][1], m[2][0], m[2][1]);
        System.out.printf(
            "  Step 3 Minor M: det([%d,%d],[%d,%d]) = (%d*%d) - (%d*%d) = %d - %d = %d%n",
            m[1][0], m[1][1], m[2][0], m[2][1],
            m[1][0], m[2][1], m[1][1], m[2][0],
            m[1][0] * m[2][1], m[1][1] * m[2][0],
            minor13
        );

        // ── Cofactor Terms ──
        // Multiply each first-row element by its minor and apply the sign pattern:
        // C11 is positive, C12 is negative, C13 is positive (checkerboard rule).
        int c11 =  m[0][0] * minor11;
        int c12 = -m[0][1] * minor12;
        int c13 =  m[0][2] * minor13;

       System.out.println();
        System.out.printf("  Cofactor C11 = (+1) x %d x %d = %d%n", m[0][0], minor11, c11);
        System.out.printf("  Cofactor C12 = (-1) x %d x %d = %d%n", m[0][1], minor12, c12);
        System.out.printf("  Cofactor C13 = (+1) x %d x %d = %d%n", m[0][2], minor13, c13);

         // -- Final Determinant --
        // Add all three signed cofactor contributions to get the determinant.
        int det = c11 + c12 + c13;
        System.out.printf("%n  det(M) = %d + (%d) + %d%n", c11, c12, c13);
        System.out.println("=".repeat(52));
        System.out.printf("  DETERMINANT = %d%n", det);
 
        // -- Singular Matrix Check --
        // A zero determinant means the matrix has no inverse (it is singular).
        if (det == 0) {
            System.out.println("  The matrix is SINGULAR -- it has no inverse.");
        }
        System.out.println("=".repeat(52));
    }
 
    // -- SECTION 5: Entry Point --------------------------------------------------
    // The program starts here. We simply pass the declared matrix into
    // the solver method and let it handle all computation and output.
    public static void main(String[] args) {
        solveDeterminant(matrix);
    }
 
}

File: MIDTERM-LAB-2/uphsd-cs-nagpaton-francescalouise/linear-algebra/assignment-01/determinant_solver.js
/**
 * =====================================================
 * Student Name    : Nagpaton, Francesca Louise May G.
 * Course          : Math 101 — Linear Algebra
 * Assignment      : Programming Assignment 1 — 3x3 Matrix Determinant Solver
 * School          : University of Perpetual Help System DALTA, Molino Campus
 * Date            : March 19, 2026
 * GitHub Repo     : https://github.com/honeyshey-ops/Prog2-9302-AY225-NAGPATON
 * Runtime         : Node.js (run with: node determinant_solver.js)
 *
 * Description:
 *   JavaScript equivalent of DeterminantSolver.java. This script computes
 *   the determinant of the same hardcoded 3x3 matrix assigned to student #23
 *   using cofactor expansion along the first row. All intermediate steps
 *   are logged to the console using console.log() so the full solution
 *   process is transparent and easy to follow.
 * =====================================================
 */

// ── SECTION 1: Matrix Declaration ───────────────────────────────────
// The assigned 3x3 matrix for Nagpaton, Francesca Louise May G. (student #23).
// It is stored as a nested JavaScript array where each inner array is one row.
// These exact values come from the assignment sheet and must not be modified.
const matrix = [
    [1, 3, 5],   // Row 1 of assigned matrix
    [2, 4, 6],   // Row 2 of assigned matrix
    [3, 5, 7]    // Row 3 of assigned matrix
];

// ── SECTION 2: Matrix Printer ────────────────────────────────────────
// Accepts the 3x3 array and prints it to the console with box-drawing
// characters so it resembles a proper mathematical matrix notation.
// Template literals are used for clean, readable string formatting.
function printMatrix(m) {
    console.log(`┌                  ┐`);
    m.forEach(row => {
        const fmt = row.map(v => v.toString().padStart(4)).join("  ");
        console.log(`│ ${fmt}   │`);
    });
    console.log(`└                  ┘`);
}

// ── SECTION 3: 2×2 Determinant Helper ───────────────────────────────
// Computes the determinant of a 2x2 matrix given four individual values.
// The four parameters represent the matrix laid out as:
//   | a  b |
//   | c  d |
// This function is called three separate times during cofactor expansion.
function computeMinor(a, b, c, d) {
    // Standard 2x2 determinant: multiply down-right diagonal, subtract down-left
    return (a * d) - (b * c);
}

// ── SECTION 4: Step-by-Step Determinant Solver ──────────────────────
// This is the main function that drives the entire solution output.
// It walks through cofactor expansion step by step and prints each part:
//   1. Displays the assigned matrix in a readable format
//   2. Computes and prints each 2x2 minor with its arithmetic breakdown
//   3. Calculates each signed cofactor term and shows the multiplication
//   4. Sums the three terms and announces the final determinant value
//   5. Prints a singular matrix warning if the determinant equals zero
function solveDeterminant(m) {
    const line = "=".repeat(52);

    // Print the problem header and display the matrix in box format
    console.log(line);
    console.log("  3x3 MATRIX DETERMINANT SOLVER");
    console.log("  Student: Nagpaton, Francesca Louise May G.");
    console.log("  Assigned Matrix:");
    console.log(line);
    printMatrix(m);
    console.log(line);
    console.log();
    console.log("Expanding along Row 1 (cofactor expansion):");
    console.log();

    // ── Step 1: Minor M₁₁ ──
    // Remove row 0 and column 0 from the matrix.
    // The remaining four elements form the sub-matrix for this minor.
    const minor11 = computeMinor(m[1][1], m[1][2], m[2][1], m[2][2]);
    console.log(
        `  Step 1 — Minor M₁₁: det([${m[1][1]},${m[1][2]}],[${m[2][1]},${m[2][2]}])` +
        ` = (${m[1][1]}×${m[2][2]}) - (${m[1][2]}×${m[2][1]})` +
        ` = ${m[1][1] * m[2][2]} - ${m[1][2] * m[2][1]} = ${minor11}`
    );

    // ── Step 2: Minor M₁₂ ──
    // Remove row 0 and column 1. Use what remains for this minor calculation.
    const minor12 = computeMinor(m[1][0], m[1][2], m[2][0], m[2][2]);
    console.log(
        `  Step 2 — Minor M₁₂: det([${m[1][0]},${m[1][2]}],[${m[2][0]},${m[2][2]}])` +
        ` = (${m[1][0]}×${m[2][2]}) - (${m[1][2]}×${m[2][0]})` +
        ` = ${m[1][0] * m[2][2]} - ${m[1][2] * m[2][0]} = ${minor12}`
    );

    // ── Step 3: Minor M₁₃ ──
    // Remove row 0 and column 2. The last set of four elements gives this minor.
    const minor13 = computeMinor(m[1][0], m[1][1], m[2][0], m[2][1]);
    console.log(
        `  Step 3 — Minor M₁₃: det([${m[1][0]},${m[1][1]}],[${m[2][0]},${m[2][1]}])` +
        ` = (${m[1][0]}×${m[2][1]}) - (${m[1][1]}×${m[2][0]})` +
        ` = ${m[1][0] * m[2][1]} - ${m[1][1] * m[2][0]} = ${minor13}`
    );

    // ── Cofactor Terms ──
    // Each first-row element is multiplied by its minor and a sign factor.
    // Signs follow the alternating pattern: +, -, + for positions 1, 2, 3.
    const c11 =  m[0][0] * minor11;
    const c12 = -m[0][1] * minor12;
    const c13 =  m[0][2] * minor13;

    console.log();
    console.log(`  Cofactor C₁₁ = (+1) × ${m[0][0]} × ${minor11} = ${c11}`);
    console.log(`  Cofactor C₁₂ = (-1) × ${m[0][1]} × ${minor12} = ${c12}`);
    console.log(`  Cofactor C₁₃ = (+1) × ${m[0][2]} × ${minor13} = ${c13}`);

    // ── Final Determinant ──
    // The determinant is the total sum of all three cofactor contributions.
    const det = c11 + c12 + c13;
    console.log();
    console.log(`  det(M) = ${c11} + (${c12}) + ${c13}`);
    console.log(line);
    console.log(`  ✓  DETERMINANT = ${det}`);

    // ── Singular Matrix Check ──
    // When the determinant is zero, the matrix is singular and has no inverse.
    if (det === 0) {
        console.log("  ⚠ The matrix is SINGULAR — it has no inverse.");
    }
    console.log(line);
}

// ── SECTION 5: Program Entry Point ──────────────────────────────────
// This line kicks off the entire program by passing the student's matrix
// into the solver function defined above.
solveDeterminant(matrix);